src/main/java/com/gooddata/executeafm/afm/PreviousPeriodMeasureDefinition.java

```diff
@@ -0,0 +1,161 @@
+     * @throws IllegalArgumentException
+     *         Thrown when {@code attributes} list is empty.
+     */
+    @JsonCreator
+    public PreviousPeriodMeasureDefinition(
+            @JsonProperty("measureIdentifier") final String measureIdentifier,
+            @JsonProperty("dateDataSets") final List<PreviousPeriodDateDataSet> dateDataSets) {
+        this.measureIdentifier = notNull(measureIdentifier, "'measureIdentifier' must not be null!");
+        this.dateDataSets = notEmpty(dateDataSets, "'dateDataSets' list must not be empty!");
+    }
+
+    /**
+     * The method is not supported by the object.
+     * Use {@link #getObjQualifiers()} instead.
+     *
+     * @return throws {@link UnsupportedOperationException}
+     *
+     * @throws UnsupportedOperationException
+     *         The exception is thrown every time the method is called.
+     * @deprecated Use {@link #getObjQualifiers()} instead.
+     */
+    @Override
+    @Deprecated
+    public ObjQualifier getObjQualifier() {
+        throw new UnsupportedOperationException("The method is not supported by the object!");
+    }
+
+    @Override
+    public Collection<ObjQualifier> getObjQualifiers() {
+        return this.dateDataSets.stream()
+                .map(dataSet -> dataSet.getDataSet())
+                .collect(Collectors.toSet());
+    }
+
+    /**
+     * The method is not supported by the object.
+     * Use {@link #withObjUriQualifiers(ObjQualifierConverter)} instead.
+     *
+     * @param uriQualifier
+     *         The URI object qualifier.
+     *
+     * @return throws {@link UnsupportedOperationException}
+     *
+     * @throws UnsupportedOperationException
+     *         The exception is thrown every time the method is called.
+     * @deprecated Use {@link #withObjUriQualifiers(ObjQualifierConverter)} instead.
+     */
+    @Override
+    @Deprecated
+    public MeasureDefinition withObjUriQualifier(final UriObjQualifier uriQualifier) {
+        throw new UnsupportedOperationException("The method is not supported by the object!");
+    }
+
+    @Override
+    public MeasureDefinition withObjUriQualifiers(final ObjQualifierConverter objQualifierConverter) {
+        notNull(objQualifierConverter, "'objQualifierConverter' must not be null!");
+        return new PreviousPeriodMeasureDefinition(measureIdentifier, copyDataSetsWithUriQualifiers(objQualifierConverter));
+    }
+
+    private List<PreviousPeriodDateDataSet> copyDataSetsWithUriQualifiers(final ObjQualifierConverter objQualifierConverter) {
+        return dateDataSets.stream()
+                .map(dataSet -> copyWithUriQualifier(dataSet, objQualifierConverter))
+                .collect(Collectors.toList());
+    }
+
+    private PreviousPeriodDateDataSet copyWithUriQualifier(final PreviousPeriodDateDataSet dataSet, final ObjQualifierConverter objQualifierConverter) {
+        if (dataSet.getDataSet() instanceof IdentifierObjQualifier) {
```

Comment:
this piece of code looks almost the same as in OverPeriod... could it be somehow shared?

Reply:
This took some work but I managed to make the generic utility used by all the measure definition implementations. See the `ObjIdentifierUtilities` and its test.